### ISSUE TITLE
Change the parent of a subnamespace

### ISSUE THREAD

**Issue by johnpemberton** [kubernetes-sigs/multi-tenancy#1045]
Hi,

I can envisage a scenario where changing the parent of a subnamespace would be really beneficial.   Is this something that HNC is likely to support?

Scenario:
 - Large organisation, lots of teams, who each look after a number of products
 - Sometimes the ownership of those products changes from one team to another 

Lets start out with a namespace tree like this:
```
- org
   - orders (fullnamespace)
      - order-management (subnamespace)
      - order-notifications (subnamespace)
   - payments (fullnamespace)
      - payment-services (subnamespace)
      - payment-notifications (subnamespace)
```

It's great that the orders and payments teams were in control of creating their own product subnamespaces without ClusterRoles, or putting in requests for sign-off for creating full namespaces from the platform team.   That's one of the things I really like about subnamespaces.

But then the org notices the duplication in that both orders and payments teams have notification services; they decide to create a notifications team, and want them to take ownership of the order-notifications and payment-notifications applications, with the end view of eliminating the duplication.  

The tree wants to look like this:

```
- org
   - orders (fullnamespace)
      - order-management (subnamespace)
   - payments (fullnamespace)
      - payment-services (subnamespace)
   - notifications (fullnamespace)
      - order-notifications (subnamespace)
      - payment-notifications (subnamespace)
```

They create a new notifications namespace, and try to move the order-notifications and payment-notifications under it as child namespaces. 

Unfortunately as the order-notifications and payment-notifications namespaces are subnamespaces, this isn't possible as you [can't change the parent of a subnamespace](https://github.com/kubernetes-sigs/multi-tenancy/blob/hnc-v0.5.1/incubator/hnc/docs/user-guide/concepts.md#subnamespaces-and-full-namespaces).   

If they were created as full namespaces, this would be possible, but that would negate the freedom that our orders and payments teams had in being able to create their own subnamespaces in the first place, and we'd be back to having the administrative burden and delay for approval/sign-off etc for full namespace creation.

I do believe this type of scenario is highly likely to happen, particularly in larger organisations, so I think the ability to reorganise the namespace hierarchy tree, whether involving full namespaces or subnamespaces would be really beneficial.

Also aware of Conway's Law :)

Thanks,
John

**Comment by johnpemberton** [comment]
@adrianludwin  

Just been playing with this again, and I think the correct sequence for moving a subnamespace within the hierarchy tree is:

- Remove the annotation:
```
kubectl annotate namespace myns hnc.x-k8s.io/subnamespaceOf-
```
- Change the parent:
```
kubectl hns set myns --parent newparent
```
- Delete the subnamespaceanchor
``` 
kubectl delete subnamespaceanchor myns -n oldparent
```
- Add the annotation back
```
kubectl annotate namespace myns hnc.x-k8s.io/subnamespaceOf=newparent
```
- Create a new subnamespaceanchor
```
cat <<EOF | kubectl apply -n newparent -f -
apiVersion: hnc.x-k8s.io/v1alpha1
kind: SubnamespaceAnchor
metadata:
  name: myns
EOF
```

If you try to delete the `subnamespaceanchor` *before* changing the parent, then you lose your namespace and you can't get rid of the offending `subnamespaceanchor`.


This script shows it working by building up a tree and successfully moving a subnamespace from one place in the hierarchy to another:

```
#!/usr/bin/env bash

function output {
    sep="\n--------------------------------------------------------\n"
    printf "$sep $1 $sep"
    sleep 2
}

output "Build a tree:"

kubectl create ns myorg

kubectl create ns team-1
kubectl create ns team-2

kubectl hns set team-1 --parent myorg
kubectl hns set team-2 --parent myorg

kubectl hns create product-1 -n team-1
kubectl hns create product-2 -n team-2

output "The initial tree looks like this:"

kubectl hns tree myorg


output "Now we want to move subnamespace for product-1 from team-1 to team 2:"


output "First we remove the hnc.x-k8s.io/subnamespaceOf annotation"
kubectl annotate namespace product-1 hnc.x-k8s.io/subnamespaceOf-

output "Next we change the parent"

kubectl hns set product-1 --parent team-2

output "Next we remove the subnamespace anchor from the old parent namespace - if you did this before you changed the parent, you'd lose the namespace"

kubectl delete subnamespaceanchors -n team-1 product-1

output "Finally we re-add the annotation, but pointing at the new parent namespace.."
kubectl annotate namespace product-1 hnc.x-k8s.io/subnamespaceOf=team-2


output "... and create a new subnamespaceanchor between the two namespaces"
cat <<EOF | kubectl apply -n team-2 -f -
apiVersion: hnc.x-k8s.io/v1alpha1
kind: SubnamespaceAnchor
metadata:
  name: product-1
EOF

output "Our tree now looks like this"
kubectl hns tree myorg

output "Now we're cleaning up"
kubectl delete subnamespaceanchor product-1 -n team-2
kubectl delete subnamespaceanchor product-2 -n team-2
kubectl delete ns team-1
kubectl delete ns team-2
kubectl delete ns myorg

```

Btw... I'm happy to pick up adding this to the docs if you haven't started yet @adrianludwin 

Thanks,
John

**Comment by adrianludwin** [comment]
Thanks for submitting the PR! I've left some comments there but in summary:
1. You should be able to recreate the "missing" anchor via `kubectl hns create`.
2. It's rare that you need to turn a full ns into a subns
3. It's better to talk about _first_ converting a subns to full ns, _then_ moving it to a new parent, and _lastly_ turning it back into a subns (though this is discouraged).

**Comment by adrianludwin** [comment]
Fixed by #1302 

/close
